### PR TITLE
libsv lua extension

### PR DESCRIPTION
**Pull request by uael**
**pending do not merge**

What's next ?
I'm trying to use the thing in some tests, i'm unable get the semver module

### CODE REVIEW

**Comment by waruqi** [comment]
@uael You need remove `semver.parse`  in `core/base/semver.lua` first. Then you can call `semver.parse` in `semver/parse.c`.

**Comment by waruqi** [comment]
👍

**Comment by uael** [comment]
I was thinking about `semver.select`
```lua
semver.select(">=1.5.0 <1.6", {"1.5.0", "1.5.1"}, {"v1.5.0", ..}, {"master", "dev"})
```
The actual prototype match and return the first version that satisfies the range and its source.
To resolve multiple and interdependent modules, the result should be:
```lua
{ versions = {"1.5.0", "1.5.1"}, tags = {"v1.5.0", ..}, branches = {} }
```

**Comment by waruqi** [comment]
But I just want to get only one best version to download the source package and I need known it's version source type(from `versions`, `tags`, `branches`?). 

You can add another interface(`semver.select_range`) to satisfy range and source. 

```lua
semver.select_range(">=1.5.0 <1.6", {"1.5.0", "1.5.1"}, {"v1.5.0", ..}, {"master", "dev"})
-- result: { versions = {"1.5.0", "1.5.1"}, tags = {"v1.5.0", ..}, branches = {} }
```

**Comment by uael** [comment]
a deps:
- b: '*'
- c: '*'

b deps:
- d: '1.5.x'

c deps:
- d: '>=1.5.0  <1.5.5'

In this case, when installing a, did you concatenate b and c requirement for d like `semver.select("1.5.x >=1.5.0  <1.5.5", ...)` ?

**Comment by waruqi** [comment]
I haven't really thought about how to deal with them. I think we can modify the `semver.select`interfaces to handle the dependencies between versions by `deps = {..}` args.

```lua
semver.select({ deps = {...}, require = ">=1.5.0 <1.6", "versions = {"1.5.0", "1.5.1"},  tags = {"v1.5.0", ..}, branches = {"master", "dev"}})
```

and return only one version info, or we can through several interfaces to implement them?

**Comment by uael** [comment]
It should be
```lua
semver.resolve({
    { 'b', '*', b_versions, b_tags, b_branches },
    { 'c', '*', c_versions, c_tags, c_branches },
    { 'd', '1.5.x >=1.5.0 <1.5.5', d_versions, d_tags, d_branches }
})
-- {{ dependency = 'b', version = '0.2', source = 'tags' }, ...}
```
Then the solvers try to satisfy dependencies by the latest versions

**Comment by uael** [comment]
I didn't understand why the travis linux build keep failing :grimacing:

File: core/src/xmake/semver/select.c [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+ * includes
+ */
+#include "prefix.h"
+
+/* //////////////////////////////////////////////////////////////////////////////////////
+ * implementation
+ */
+
+// satisfies wrapper
+tb_int_t xm_semver_select(lua_State* lua)
```
Comment (waruqi) [inline core/src/xmake/semver/select.c]: 👍

**Comment by waruqi** [comment]
@uael  
> semver.resolve({
    { 'b', '*', b_versions, b_tags, b_branches },
    { 'c', '*', c_versions, c_tags, c_branches },
    { 'd', '1.5.x >=1.5.0 <1.5.5', d_versions, d_tags, d_branches }
})

Great!  👍

**Comment by waruqi** [comment]
@uael 
> I didn't understand why the travis linux build keep failing

I will look at it tomorrow.

**Comment by waruqi** [comment]
@uael You can merge the commit 7d7d841e4731d0f22ae3ba4427fed97dc7d08803 to get more verbose error info when building fails on travis-ci

**Comment by uael** [comment]
@waruqi Thanks :+1:

**Comment by uael** [comment]
Before writing the solver i will lookup what's already exist in well designed package managers

**Comment by uael** [comment]
So after some research I've find how to implement this, I'm going to have a lot more time in a week but I will keep trying to work on it before.
@waruki How to merge a single commit ? (for 7d7d841)

**Comment by waruqi** [comment]
@uael You can update and merge all commits in `repo` branch. 

```bash
$ git fetch https://github.com/tboox/xmake.git repo:tmp
$ git rebase tmp
$ git push origin repo
$ git branch -d tmp
```

**Comment by waruqi** [comment]
@uael  You need to add these files to `core/src/xmake/makefile` 

```
make: *** [build] Error 2
error xmake
../xmake/libxmaker.a(machine.o):(.rodata+0x8): undefined reference to `xm_semver_parse'
../xmake/libxmaker.a(machine.o):(.rodata+0x18): undefined reference to `xm_semver_satisfies'
../xmake/libxmaker.a(machine.o):(.rodata+0x28): undefined reference to `xm_semver_select'
collect2: error: ld returned 1 exit status
Build Fail
```

**Comment by waruqi** [comment]
@uael The changed commits(61) and files(96) seem be too much, I do not know whether there is a problem, can you new pr and try it again?

**Comment by uael** [comment]
@waruqi 
> The changed commits(61) and files(96) seem be too much, I do not know whether there is a problem, can you new pr and try it again?
Haha wtf, yep

**Pull request closed without merging**